src/diffusers/pipelines/unidiffuser/modeling_text_decoder.py

```diff
@@ -0,0 +1,188 @@
+from typing import Optional
+
+import numpy as np
+import torch
+from torch import nn
+from transformers import GPT2LMHeadModel, GPT2Tokenizer
+
+from ...configuration_utils import ConfigMixin, register_to_config
+from ...models import ModelMixin
+
+
+# Modified from ClipCaptionModel in https://github.com/thu-ml/unidiffuser/blob/main/libs/caption_decoder.py
+class UniDiffuserTextDecoder(ModelMixin, ConfigMixin):
+    @register_to_config
+    def __init__(
+        self,
+        tokenizer: GPT2Tokenizer,
+        text_decoder: GPT2LMHeadModel,
```

Comment:
Can we try to seperate the tokenizer and text decoder here.

`diffusers` should be able to load the tokenizer out of the box, you just have to define it in the pipeline:

````
def __init__(
    ...
    tokenizer,
):
```

We can then also add the `"<|EOS|>"` token directly in the init of the pipeline, e.g. here: https://github.com/huggingface/diffusers/pull/2963/files#r1159526676

Reply:
Also we cannot pass the `text_decoder` here at init as this would prevent us to be able to use `from_pretained(...)` of the model class. Could you maybe try to follow the design as done here: https://github.com/huggingface/diffusers/blob/a9477bbdac9b8b1755e0383e5198e4baa62678b1/src/diffusers/pipelines/spectrogram_diffusion/continous_encoder.py#L29

See how we import blocks from `transformers` to design our own new model. I think here you could just do the following:

```py
def __init__(
    self,
    num_layers=12,
    ...
):
    config = GPT2Config(...take all the config params from init)
    self.text_decoder = GPT2LMHeadModel(config)
```

We then design a new checkpoint architecture for the `UniDiffusersTextDecoder` and upload pretrained weights for it

Reply:
Changed the design of `__init__` following the example: removed `tokenizer` and `text_decoder` args, added GPT2 config args.